### PR TITLE
Basic taproot BIP32 derivation & keyspend support

### PR DESCRIPTION
**Pull request by kdmukai**
* Adds basic support for the BIP-371 Taproot psbt field `PSBT_IN_TAP_BIP32_DERIVATION` (`0x16`) to properly parse bip32 derivations in Taproot psbts.
* Only supports the root ("internal") key; does nothing if the psbt is expecting to be signed by a key within a leaf.
* Also adds `PSBT_IN_TAP_INTERNAL_KEY` (`0x17`), though I believe it won't be needed until keys within leaves are supported (internal key will match the `PSBT_IN_TAP_BIP32_DERIVATION` pubkey when `num_leaf_hashes` is zero).
* Adds the matching `PSBT_OUT_TAP_INTERNAL_KEY` (`0x05`) and `PSBT_OUT_TAP_BIP32_DERIVATION ` (`0x07`) fields to outputs.
* Stores the data in `InputScope.taproot_bip32_derivations` which is a tuple: `([leaf_hashes], DerivationPath)`. Same for `OutputScope.taproot_bip32_derivations`.
* Stores the internal key in `InputScope.taproot_internal_key` and `OutputScope.taproot_internal_key`.
* Signing is backwards-compatible with "legacy" Taproot implementation for psbts that do not have the `PSBT_IN_TAP_BIP32_DERIVATION` field (i.e. they populate `0x06`/`InputScope.bip32_derivations`).
* Updates Taproot psbt signing to expect X-only pubkeys.
* Updates test cases to test that `taproot_bip32_derivations` were populated and adds fixes so that the X-only pubkeys expectation passes.

---

I will also try to make a PR to update Specter Desktop to output the `0x16` and `0x17` BIP-371 fields. Because of the backwards-compatibility here, a Specter-DIY release can come first and then update Specter Desktop with these fields.

---

TODO:
Update `test_tapoot.TaprootTest.test_sign_verify` to use a psbt that includes the `PSBT_IN_TAP_BIP32_DERIVATION` field rather than the "legacy" psbt approach.

TODO:
`test_psbtview.PSBTViewTest` is failing on `test_sign`:
```
FAIL: test_sign (tests.test_psbtview.PSBTTest)
Test if we can sign psbtview and get the same as from signing psbt
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/kdmukai/dev/embit/tests/tests/test_psbtview.py", line 113, in test_sign
    self.assertEqual(ser.getvalue(), psbt.serialize())
AssertionError: b'psb[844 chars]0\x00"\x02\x02\xa9\x81\xb3?~\xb9*\n\xcf77P\x92[890 chars]\x00' != b'psb[844 chars]0\x00\x01\x01\x1f\x80\x96\x98\x00\x00\x00\x00\[2246 chars]\x00'
```
This failure pre-dates these changes so something was either already broken or something is off with my local setup.

### CODE REVIEW

**Comment by kdmukai** [comment]
FYI - still making improvements and tweaks so this PR should be considered DRAFT (but I don't seem to have permissions to revert the PR status to DRAFT).

**Review by stepansnigirev — CHANGES_REQUESTED** [review]
PSBT parsing should not fail if the library can't sign - maybe it's not used for signing at all. You can raise when signing, but I'd better ignore inputs we don't know how to sign.
For example, when you try to sign PSBT transaction with a wrong key it doesn't raise an error, it just returns 0 as a number of added signatures. I think for tap leafs we should do the same.

File: tests/tests/test_psbtview.py [hunk 1]
```diff
@@ -69,6 +69,7 @@ def test_sign(self):
         """Test if we can sign psbtview and get the same as from signing psbt"""
         for compress in [True, False]:
             for b64 in PSBTS:
+                print(b64)
```
[the remark this answers is not in the document]
Comment (kdmukai) [reply tests/tests/test_psbtview.py]: Leftover artifact of my debugging. Removed.

File: src/embit/psbt.py [hunk 2]
```diff
@@ -133,6 +134,11 @@ def __init__(self, unknown: dict = {}, vin=None, compress=CompressMode.KEEP_ALL)
         self.redeem_script = None
         self.witness_script = None
         self.bip32_derivations = OrderedDict()
+
+        # tuples of (num_leaf_hashes, [leaf_hashes], DerivationPath)
```
[the remark this answers is not in the document]
Comment (kdmukai) [reply src/embit/psbt.py]: Just being too literal in my thinking! Removed with matching changes elsewhere.

File: src/embit/psbt.py [hunk 3]
```diff
@@ -303,12 +317,33 @@ def read_value(self, stream, k):
...
+        elif k[0] == 0x16:
+            pub = ec.PublicKey.from_xonly(k[1:])
+            if pub in self.taproot_bip32_derivations:
+                raise PSBTError("Duplicated derivation path")
+            else:
+                # Field begins with the number of leaf hashes; for now only support the
+                # internal key where there are no leaf hashes.
+                # TODO: Support keysigns from leaves within the taptree.
+                if v[0] != 0:
+                    raise PSBTError("Signing for public keys in leaves not yet implemented")
```
[the remark this answers is not in the document]
Comment (kdmukai) [reply src/embit/psbt.py]: Removed these exceptions for parsing and for signing.

**Review by stepansnigirev — APPROVED** [review]

**Pull request merged**